Automattic/wp-calypso
# composite-checkout: Enable 3DS for Stripe payment method

#### Changes proposed in this Pull Request

When paying with a credit card that requires 3DS authentication, Stripe's library will display a modal to perform the authentication and then tell the page if the auth completed or failed.

This PR hooks up the response from the modal to the rest of the form, so that `onPaymentComplete` will be called when the authentication completes for both new and existing cards.

Additionally, this fixes a regression introduced in #38759 which caused the stripe form to never finish loading. This happened because the payment methods were being recreated twice; the first time the stripe payment method went and fetched the stripe configuration, but then the second time the configuration was no longer present and the hook did not try fetching it again. The cause of this double-rendering was that the `useStoredCards` hook was calling setState twice; once to set `isLoading` to false and again to save the stored cards. This PR combines these two states into a reducer so that they operate together and only cause a single render.

#### Testing instructions

- Sandbox the store.
- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`.
- Add a plan to your cart and visit checkout.
- Enter a 3DS stripe test card, like `4000000000003220`.
- Complete checkout and press Pay.
- When the auth dialog appears, confirm it, and verify that the payment is successful and that the product is added to your site.
- Repeat the above process but refuse the auth dialog and verify that the payment fails and the product is not added.
- Repeat the above process with an existing card that requires 3DS (eg: the one you just added by completing the above transaction).
